docker/django/Dockerfile

```diff
@@ -1,18 +1,45 @@
-FROM python:3.6
+ARG PYTHON_VERSION=3.6
+ARG IMAGE_VARIANT=slim
 
-ENV PYTHONUNBUFFERED 0
+# ==============================
+FROM helsinkitest/python:${PYTHON_VERSION}-${IMAGE_VARIANT} AS base_stage
+# ==============================
 
-RUN apt-get update \
-    && apt-get install --no-install-recommends -y gdal-bin python-gdal python3-gdal
+ENV PYTHONDONTWRITEBYTECODE true
+ENV PYTHONUNBUFFERED true
 
-COPY requirements.txt /usr/src/
+# Add tini init system https://github.com/krallin/tini
+ENV TINI_VERSION v0.18.0
+ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /tini
+RUN chmod +x /tini
 
-RUN pip install --upgrade pip \
-    && pip install pip-tools \
-    && pip install --src /usr/src -r /usr/src/requirements.txt
+COPY --chown=appuser:appuser ./docker/django/docker-entrypoint.sh /app/
+ENTRYPOINT ["/tini", "--", "/app/docker/django/docker-entrypoint.sh"]
 
-RUN mkdir /code
+EXPOSE 8000
 
-WORKDIR /code
+# ==============================
+FROM base_stage AS development
+# ==============================
 
-COPY . /code
+RUN apt-install.sh gdal-bin build-essential
+
+COPY --chown=appuser:appuser requirements.txt /app/
```

Comment:
you could move this command to the `base_stage` to avoid repeating it.

Reply:
as well as avoid repeating `pip install --no-cache-dir -r /app/requirements.txt`

Reply:
I've intentionally written the `development` stage like above. I could write this in the `base_stage`:

```Dockerfile
COPY --chown=appuser:appuser requirements.txt /app/

RUN apt-install.sh build-essential gdal-bin \
    && pip install --no-cache-dir -r /app/requirements.txt \
    && apt-cleanup.sh build-essential gdal-bin
``` 

But then, if `requirements.txt` changes, `COPY` directive and anything after that invalidates. Meaning that installation of `build-essential gdal-bin` has to be done again.

At the moment, with the way `development` stage is written, if `requirements.txt` changes:

```Dockerfile
RUN apt-install.sh gdal-bin build-essential
```

remains valid. It's IMO important to keep the `development` stage build as fast as possible, as the need to rebuild images in normal dev work is frequent enough.

Reply:
well, I guess it boils down to preference. on the one hand you have faster builds, if you install some libraries in a layer before `COPY requirements* ...` (but with a relatively new computer the time you save this way is very minimal). on the other hand you have cleaner Dockerfile, which has less repeating lines (but I guess it is not THAT messy if you have a couple of duplicate lines). 🤔 at least I don't know of clear technical advantages of either one of the approaches (although it might exist and someone with more experience would tell us about it 😅 ) 

I guess, we can leave the final decision for you. you can keep whichever you prefer. 🙂

Reply:
I changed my mind, I think you're right on the limited time saving opportunity versus the benefit of a DRY-er version of Dockerfile 👍 Done in 4489697